src/kessler.F

```diff
@@ -736,80 +736,98 @@ subroutine fallout_GPU(dt,train,ruh,rvh,zh,mh,mf,rain,prate,rr,rho,   &
       real, dimension(ib:ie,jb:je,kb:ke) :: rr,rho,q3d,vq
 
       integer :: i,j,k,n,nr,nrk
-      real, dimension(-1:nk+3):: rq
+      real, dimension(nj,-1:nk+3):: rq
       real, dimension(0:nk) :: ffk
       real, dimension(nk) :: qtmp
 
-      integer,dimension(ni) :: nfall !made into an array by SM
-      real,dimension(ni) :: dtfall !made into an array by SM
+      integer,dimension(ni,nj) :: nfall 
+      integer :: ll
+      real,dimension(ni,nj) :: dtfall
       real :: crmax
 
       double precision :: tem
       double precision, dimension(nj) :: bud
 
 !--------------------------------------------------------------------
 
-    !$acc data create(bud,nfall,dtfall)
-    !$acc parallel default(present) private(i,j,k,crmax,rq,n,nr) 
-    !$acc loop gang vector 
+    !$acc data create(bud,nfall,dtfall,rq)
+    !$acc parallel loop gang vector  default(present) private(j) 
     !$omp parallel do default(shared) private(i,j,k,crmax,nfall,rq,n,dtfall,nr)
     do j=1,nj
-    bud(j)=0.0d0
-    !$acc loop gang vector
+      bud(j)=0.0d0
+    enddo
+    !$acc parallel loop gang vector collapse(3) default(present) private(i,j,k,n,nr) 
+    do j=1,nj
     do i=1,ni
-      crmax = 0.0
-      !$acc loop gang vector reduction(max:crmax)
       do k=1,nk
+        if (k==1) crmax = 0.0
         crmax = max( crmax , vq(i,j,k)*dt*rdz*mh(i,j,k) )
       enddo
-      nfall(i) = max( 1 , int(crmax+1.0) )
-      dtfall(i)=dt/nfall(i)
+      nfall(i,j) = max( 1 , int(crmax+1.0) )
+      dtfall(i,j)=dt/nfall(i,j)
+    enddo
     enddo
       ! cm1r17:  following code is needed if nfall is large.
       !  - below edge of falling precip, set fallspeed to
       !    value at gridpoint above:
-    !$acc loop gang vector collapse(2)
+    !$acc parallel loop gang vector default(present) private(i,j,k,n) collapse(3)
+    do k=(nk-1),1,-1
+    do j=1,nj
     do i=1,ni
-      do k=(nk-1),1,-1
         if( q3d(i,j,k).lt.1.0e-8 ) vq(i,j,k) = vq(i,j,k+1)
       enddo
     enddo
-      !!dtfall=dt/nfall(i) moved to up above
-    !$acc loop collapse(3) gang vector
+    enddo
+    !$acc parallel default(present) private(i,j,k,n)
+    !$acc loop gang vector collapse(3) 
+    do k=1,nk
+    do j=1,nj
     do i=1,ni
-      do n=1,nfall(i)
-        do k=1,nk
-         rq(k)=rho(i,j,k)*vq(i,j,k)*max(0.0,q3d(i,j,k))
-          rq(nk+1)=0.0 ! Added here from below by SM
-        enddo
+         rq(j,k)=rho(i,j,k)*vq(i,j,k)*max(0.0,q3d(i,j,k))
+         rq(j,nk+1)=0.0
       enddo
     enddo
-    !$acc loop collapse(3) gang vector
+    enddo
+    !$acc end parallel
+
+    !$acc parallel default(present) private(i,j,k,n)
+    !$acc loop gang vector collapse(3) 
+    do k=1,nk
+    do j=1,nj
     do i=1,ni
```

Comment:
What is the i-loop doing  here?  Is this a bug that needs to be discussed with George?

Reply:
The latest commit should address this issue.